scalalib/src/mill/scalalib/JavaModule.scala

```diff
@@ -1380,6 +1472,90 @@ trait JavaModule
     Some((JvmBuildTarget.dataKind, bspJvmBuildTargetTask()))
   }
+
+  // Return the directory containing the current call graph analysis results, and previous one too if it exists
+  def callGraphAnalysis: T[(Path, Option[Path])] = Task(persistent = true) {
+    os.remove.all(Task.dest / "previous")
+    if (os.exists(Task.dest / "current"))
+      os.move.over(Task.dest / "current", Task.dest / "previous")
+    val debugEnabled = Task.log.debugEnabled
+    mill.codesig.CodeSig
+      .compute(
+        classFiles = os.walk(compile().classes.path).filter(_.ext == "class"),
+        upstreamClasspath = compileClasspath().toSeq.map(_.path),
+        ignoreCall = { (callSiteOpt, calledSig) =>
+          // We can ignore all calls to methods that look like Targets when traversing
+          // the call graph. We can do this because we assume `def` Targets are pure,
+          // and so any changes in their behavior will be picked up by the runtime build
+          // graph evaluator without needing to be accounted for in the post-compile
+          // bytecode callgraph analysis.
+          def isSimpleTarget(desc: mill.codesig.JvmModel.Desc) =
+            (desc.ret.pretty == classOf[mill.define.Target[?]].getName ||
+              desc.ret.pretty == classOf[mill.define.Worker[?]].getName) &&
+              desc.args.isEmpty
+
+          // We avoid ignoring method calls that are simple trait forwarders, because
+          // we need the trait forwarders calls to be counted in order to wire up the
+          // method definition that a Target is associated with during evaluation
+          // (e.g. `myModuleObject.myTarget`) with its implementation that may be defined
+          // somewhere else (e.g. `trait MyModuleTrait{ def myTarget }`). Only that one
+          // step is necessary, after that the runtime build graph invalidation logic can
+          // take over
+          def isForwarderCallsiteOrLambda =
+            callSiteOpt.nonEmpty && {
+              val callSiteSig = callSiteOpt.get.sig
+
+              (callSiteSig.name == (calledSig.name + "$") &&
+                callSiteSig.static &&
+                callSiteSig.desc.args.size == 1)
+              || (
+                // In Scala 3, lambdas are implemented by private instance methods,
+                // not static methods, so they fall through the crack of "isSimpleTarget".
+                // Here make the assumption that a zero-arg lambda called from a simpleTarget,
+                // should in fact be tracked. e.g. see `integration.invalidation[codesig-hello]`,
+                // where the body of the `def foo` target is a zero-arg lambda i.e. the argument
+                // of `Cacher.cachedTarget`.
+                // To be more precise I think ideally we should capture more information in the signature
+                isSimpleTarget(callSiteSig.desc) && calledSig.name.contains("$anonfun")
+              )
+            }
```

Comment:
All the other special cases except this one are heuristics unique to `MillBuildRootModule`, and so probably should live there. We can have a `def codeSigIgnoreCall` method that gets fed into `CodeSig.compute`, with a default implementation for `JavaModule` and a more detailed override for `MillBuildRootModule`